src/colorize.cr

```diff
@@ -214,10 +226,33 @@ module Colorize
       {red, green, blue}.join(';', io, &.to_s io)
     end
   end
+
+  # A text decoration.
+  #
+  # Note that not all decorations are supported in all terminals.
+  # When a decoration is not supported, the text won't have any decoration.
+  @[Flags]
+  enum Mode
+    # Makes the text bold.
+    Bold = 1
+    # Makes the text color bright.
+    Bright = 1
+    # Dims the text color, the opposite of `Bold` and `Bright`.
+    Dim
+    # Underlines the text.
+    Underline
+    # Makes the text blink slowly.
+    Blink
+    # Swaps the foreground and background colors.
+    Reverse
+    # Makes the text invisible.
+    Hidden
+  end
 end
 
+# A colorize object colors and decorations can be applied to.
 struct Colorize::Object(T)
-  private MODE_DEFAULT   = '0'
+  private MODE_NONE      = '0'
```

Comment:
`MODE_*` private constants can be changed to a private enum annotated with `@[Flags]`, which will bring `Mode::None`.

Reply:
I would like to not make these constants an enum because enums can't contain `Char`s so they would need to be integers and that would mean a bunch of `to_s`s which I would like to avoid.

Reply:
The performance difference isn't significant. I think this enum should be exposed and documented, in order to know available modes.

```cr
require "benchmark"

io = File.open File::DEVNULL, "w"

enum E
  A
  B
  C
end

E_A = '1'
E_B = '2'
E_C = '3'

Benchmark.ips do |x|
  x.report("enum") do
    io << E::A
    io << E::B
    io << E::C
  end

  x.report("char") do
    io << E_A
    io << E_B
    io << E_C
  end
end
```
```
enum  42.39M ( 23.59ns) (± 6.34%)  0.0B/op   1.72× slower
char  73.07M ( 13.69ns) (± 4.67%)  0.0B/op        fastest
```

Reply:
This enum is actually exposed and documented here: [line 235](https://github.com/crystal-lang/crystal/pull/7690/files#diff-9fc1ac5e1582956c996fd04dde1f3049R235). These constants are basically part of this enum and just there because the `Mode` enum is a flags enum and not in normal order which is required here: [line 430](https://github.com/crystal-lang/crystal/pull/7690/files#diff-9fc1ac5e1582956c996fd04dde1f3049R430).

Reply:
The enum already exists, why not directly use it?

Reply:
Because the `Mode` enum can't be used directly. The `Mode` enum is a flags enum and is thus not in normal order from 1 to 7 but that is required in [line 430](https://github.com/crystal-lang/crystal/pull/7690/files#diff-9fc1ac5e1582956c996fd04dde1f3049R430). The constants however are in the correct normal number order from 1 to 7.

Reply:
Sorry I don't understand, what is line 430?

Reply:
The links broke because of a commit, the line is at [404](https://github.com/crystal-lang/crystal/pull/7690/files#diff-9fc1ac5e1582956c996fd04dde1f3049R404) now.

Reply:
Maybe `Bright` can be removed, it has the same effect has bold ([reference](https://en.wikipedia.org/wiki/ANSI_escape_code))

Anyway you can do
```cr
{% for mode in Mode.constants %}
  {% if !["All", "None", "Bright"].includes? mode.stringify %}\
            Mode::{{mode.id}}
  {% end %}      
{% end %}  
```
Then drop private constants.

Reply:
If you really want chars, a counter can be also used

Reply:
I think bright and bold both exist because on some terminals they actually make the text **bold** and on some terminals (like mine) it just brightens up the color. So you can choose one based on how they act in your terminal.

So we are talking about this, right?
https://github.com/crystal-lang/crystal/blob/e3b84b632e7012d079fe57a1b5df4b7f50593b71/src/colorize.cr#L399-L408
Here it iterates through the mode array (`%w(bold dim underline blink reverse hidden)`) and generates an if check for each mode and if the local variable `mode` (not the one from the macro `for`) includes for example `Mode::Blink`, a semicolon is appended and then `MODE_BLINK`'s value is appended, `'5'`. So everything fine. Now when I however use the value from the `Mode` enum, I would append `8` which is wrong and would result in an incorrect ANSI escape code.
Or am I misunderstanding something?

Reply:
The `Mode` enum is incorrect. After looking at [the VT100 specification](https://vt100.net/docs/vt100-ug/chapter3.html),  there is table:
```
0 | Attributes off
1 | Bold or increased intensity
4 | Underscore
5 | Blink
7 | Negative (reverse) image
```

The enum should be
```cr
enum Mode
  Normal     = 0
  Bold       = 1
  Underscore = 4
  Blink      = 5
  Reverse    = 7
  Conceal    = 8
end
```
This enum maps to incorrect values, and can't be used outside `Colorize`.

Reply:
According to the [VT510 specification](https://vt100.net/docs/vt510-rm/chapter12.html), more values can be added to the enum.

Reply:
The values of the `Mode` enum don't matter. It's just a flags enum so you can have multiple modes. The values of the private constants matter and they are correct according to https://en.wikipedia.org/wiki/ANSI_escape_code#SGR_parameters.

Reply:
Why don't make them correct, and only use this very enum?
The private constants won't be needed then.

Reply:
As I said the `Mode` enum needs to be a flags enum so that you can have multiple modes at once for one text. Otherwise you couldn't make a text both underlined and bold.